File: Euclid.Storage/Database/Access/RoomDao.cs
using Euclid.Storage.Database.Data;
using NHibernate.Criterion;
using NHibernate.Linq;
using System.Collections.Generic;
using System.Linq;

namespace Euclid.Storage.Database.Access
{
    public class RoomDao
    {
        public static List<RoomData> SearchRooms(string query, int roomLimit = 50)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                RoomData roomDataAlias = null;
                PlayerData playerDataAlias = null;

                /*
                var test = session.QueryOver<RoomData>(() => roomDataAlias)
                    .JoinQueryOver<PlayerData>(() => roomDataAlias.OwnerData, () => playerDataAlias)
                        .Where(() => playerDataAlias.Id == roomDataAlias.OwnerId)
                    .Where(
                        Restrictions.On(() => roomDataAlias.Name).IsInsensitiveLike(query, MatchMode.Anywhere) ||
                        Restrictions.On(() => playerDataAlias.Name).IsInsensitiveLike(query, MatchMode.Anywhere))
                    //.Select(Projections.Entity(() => playerDataAlias))
                    //.Select(x => x.OwnerData);
                    .List<PlayerData>() as List<PlayerData>;*/

                return session.QueryOver<RoomData>(() => roomDataAlias)
                    .JoinQueryOver<PlayerData>(() => roomDataAlias.OwnerData, () => playerDataAlias)
                        .Where(() => playerDataAlias.Id == roomDataAlias.OwnerId)
                    .Where(
                        Restrictions.On(() => roomDataAlias.Name).IsInsensitiveLike(query, MatchMode.Anywhere) ||
                        Restrictions.On(() => playerDataAlias.Name).IsInsensitiveLike(query, MatchMode.Anywhere))
                    .OrderBy(() => roomDataAlias.UsersNow).Desc
                    .OrderBy(() => roomDataAlias.Rating).Desc
                    .Take(roomLimit)
                    .List<RoomData>() as List<RoomData>;
            }
        }

        /// <summary>
        /// Delete room
        /// </summary>
        /// <param name="roomId"></param>
        public static void DeleteRoom(int roomId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                session.Query<RoomData>().Where(x => x.Id == roomId).Delete();
            }
        }

        /// <summary>
        /// Get the list of users' own rooms
        /// </summary>
        public static List<RoomData> GetPopularFlats(int page, int resultsLimit = 11)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                RoomData roomDataAlias = null;
                PlayerData playerDataAlias = null;

                return session.QueryOver<RoomData>(() => roomDataAlias)
                    .JoinQueryOver<PlayerData>(() => roomDataAlias.OwnerData, () => playerDataAlias)
                        .Where(() => playerDataAlias.Id == roomDataAlias.OwnerId)
                    //.Where(() => roomDataAlias.CategoryId == categoryId)
                    .OrderBy(() => roomDataAlias.UsersNow).Desc
                    .Skip(page * resultsLimit)
                    //.OrderBy(() => roomDataAlias.Rating).Desc
                    .Take(resultsLimit)
                    .List() as List<RoomData>;
            }
        }
        /// <summary>
        /// Get the list of users' own rooms
        /// </summary>
        public static List<RoomData> GetPublicFlats(int categoryId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                RoomData roomDataAlias = null;

                /*
                var test = session.QueryOver<RoomData>(() => roomDataAlias)
                    .JoinQueryOver<PlayerData>(() => roomDataAlias.OwnerData, () => playerDataAlias)
                        .Where(() => playerDataAlias.Id == roomDataAlias.OwnerId)
                    .Where(
                        Restrictions.On(() => roomDataAlias.Name).IsInsensitiveLike(query, MatchMode.Anywhere) ||
                        Restrictions.On(() => playerDataAlias.Name).IsInsensitiveLike(query, MatchMode.Anywhere))
                    //.Select(Projections.Entity(() => playerDataAlias))
                    //.Select(x => x.OwnerData);
                    .List<PlayerData>() as List<PlayerData>;*/

                return session.QueryOver<RoomData>(() => roomDataAlias)
                    .Where(() => roomDataAlias.OwnerId == 0)
                    .Where(() => roomDataAlias.CategoryId == categoryId)
                    .OrderBy(() => roomDataAlias.UsersNow).Desc
                    .OrderBy(() => roomDataAlias.Rating).Desc
                    .List() as List<RoomData>;
            }
        }

        public static List<RoomData> GetPublicFlats()
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                RoomData roomDataAlias = null;

                return session.QueryOver<RoomData>(() => roomDataAlias)
                    .Where(() => roomDataAlias.OwnerId == 0)
                    .OrderBy(() => roomDataAlias.UsersNow).Desc
                    .OrderBy(() => roomDataAlias.Rating).Desc
                    .List() as List<RoomData>;
            }
        }

        /// <summary>
        /// Get the list of users' own rooms
        /// </summary>
        public static List<RoomData> GetUserRooms(int userId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                RoomData roomDataAlias = null;

                return session.QueryOver<RoomData>(() => roomDataAlias)
                    .Where(() => roomDataAlias.OwnerId == userId)
                    .OrderBy(() => roomDataAlias.UsersNow).Desc
                    .OrderBy(() => roomDataAlias.Rating).Desc
                    .List() as List<RoomData>;
            }
        }

        /// <summary>
        /// Count the rooms the user has.
        /// </summary>
        public static int CountUserRooms(int userId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<RoomData>().Where(x => x.OwnerId == userId).RowCount();
            }
        }

        /// <summary>
        /// Get the room model data
        /// </summary>
        /// <returns></returns>
        public static List<RoomModelData> GetModels()
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<RoomModelData>().List() as List<RoomModelData>;
            }
        }

        /// <summary>
        /// Get data for room
        /// </summary>
        public static RoomData GetRoomData(int roomId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<RoomData>().Where(x => x.Id == roomId).Take(1).SingleOrDefault();
            }
        }

        /// <summary>
        /// Save room data
        /// </summary>
        public static void SaveRoom(RoomData data)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        session.Update(data);
                        transaction.Commit();
                        session.Refresh(data);
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }

        /// <summary>
        /// New room data
        /// </summary>
        public static void NewRoom(RoomData data)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        session.Save(data);
                        transaction.Commit();
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }

        /// <summary>
        /// Reset all visitors
        /// </summary>
        public static void ResetVisitorCounts()
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                session.Query<RoomData>().Where(x => x.UsersNow > 0 || x.UsersNow < 0).Update(x => new RoomData { UsersNow = 0 });
            }
        }

        /// <summary>
        /// Update users count
        /// </summary>
        public static void SetVisitorCount(int roomId, int visitorsNow)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                session.Query<RoomData>().Where(x => x.Id == roomId).Update(x => new RoomData { UsersNow = visitorsNow });
            }
        }
    }
}

File: euclid-storage/src/main/java/euclid/storage/database/access/RoomDao.java
package euclid.storage.database.access;

import euclid.storage.database.SessionFactoryBuilder;
import euclid.storage.database.data.PlayerData;
import euclid.storage.database.data.RoomData;
import euclid.storage.database.data.RoomModelData;
import org.hibernate.Session;
import org.hibernate.Transaction;
import org.hibernate.query.MutationQuery;
import org.hibernate.query.Query;

import java.util.List;

public class RoomDao {

    public static List<RoomData> searchRooms(String query, int roomLimit) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<RoomData> q = session.createQuery(
                    "FROM RoomData r JOIN r.ownerData o WHERE o.id = r.ownerId " +
                    "AND (LOWER(r.name) LIKE :query OR LOWER(o.name) LIKE :query) " +
                    "ORDER BY r.usersNow DESC, r.rating DESC", RoomData.class);
            q.setParameter("query", "%" + query + "%");
            q.setMaxResults(roomLimit);
            return q.getResultList();
        }
    }

    public static void deleteRoom(int roomId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                MutationQuery query = session.createMutationQuery(
                        "DELETE FROM RoomData r WHERE r.id = :id");
                query.setParameter("id", roomId);
                query.executeUpdate();
                transaction.commit();
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static List<RoomData> getPopularFlats(int page, int resultsLimit) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<RoomData> q = session.createQuery(
                    "FROM RoomData r LEFT JOIN FETCH r.ownerData o WHERE o.id = r.ownerId " +
                    "ORDER BY r.usersNow DESC", RoomData.class);
            q.setFirstResult(page * resultsLimit);
            q.setMaxResults(resultsLimit);
            return q.getResultList();
        }
    }

    public static List<RoomData> getPublicFlats(int categoryId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<RoomData> q = session.createQuery(
                    "FROM RoomData r WHERE r.ownerId = 0 AND r.categoryId = :categoryId " +
                    "ORDER BY r.usersNow DESC, r.rating DESC", RoomData.class);
            q.setParameter("categoryId", categoryId);
            return q.getResultList();
        }
    }

    public static List<RoomData> getPublicFlats() {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<RoomData> q = session.createQuery(
                    "FROM RoomData r WHERE r.ownerId = 0 " +
                    "ORDER BY r.usersNow DESC, r.rating DESC", RoomData.class);
            return q.getResultList();
        }
    }

    public static List<RoomData> getUserRooms(int userId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<RoomData> q = session.createQuery(
                    "FROM RoomData r WHERE r.ownerId = :userId " +
                    "ORDER BY r.usersNow DESC, r.rating DESC", RoomData.class);
            q.setParameter("userId", userId);
            return q.getResultList();
        }
    }

    public static int countUserRooms(int userId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<Long> q = session.createQuery(
                    "SELECT COUNT(r) FROM RoomData r WHERE r.ownerId = :userId", Long.class);
            q.setParameter("userId", userId);
            return q.getSingleResult().intValue();
        }
    }

    public static List<RoomModelData> getModels() {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<RoomModelData> q = session.createQuery("FROM RoomModelData", RoomModelData.class);
            return q.getResultList();
        }
    }

    public static RoomData getRoomData(int roomId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            return session.find(RoomData.class, roomId);
        }
    }

    public static void saveRoom(RoomData data) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                session.update(data);
                transaction.commit();
                session.refresh(data);
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static void newRoom(RoomData data) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                session.persist(data);
                transaction.commit();
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static void resetVisitorCounts() {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Transaction tx = session.beginTransaction();
            try {
                MutationQuery query = session.createMutationQuery(
                        "UPDATE RoomData r SET r.usersNow = 0 WHERE r.usersNow > 0 OR r.usersNow < 0");
                query.executeUpdate();
                tx.commit();
            } catch (Exception e) {
                tx.rollback();
                throw e;
            }
        }
    }

    public static void setVisitorCount(int roomId, int visitorsNow) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Transaction tx = session.beginTransaction();
            try {
                MutationQuery query = session.createMutationQuery(
                        "UPDATE RoomData r SET r.usersNow = :count WHERE r.id = :id");
                query.setParameter("count", visitorsNow);
                query.setParameter("id", roomId);
                query.executeUpdate();
                tx.commit();
            } catch (Exception e) {
                tx.rollback();
                throw e;
            }
        }
    }
}
